### PR TITLE
RAINCATCH-1265 - keycloak offline

### PR DESCRIPTION
**Pull request by wtrocki**
## Motivation
 
Allow to use application when offline (and keycloak auth is enabled)

- [x] Refactor keycloak implementation
- [x] Detect offline state
- [ ] WIP Cover offline login/non auth cases (separate page needed)
- [ ] WIP Restart angular router when keycloak is reinitialized

### CODE REVIEW

File: packages/angularjs-auth-keycloak/lib/keycloakAuth.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+KeycloakAuth.prototype.hasResourceRole = function() {
+  if (this.keycloakLib) {
+    this.keycloakLib.hasResourceRole(arguments);
+  }
+}
+
+KeycloakAuth.prototype.getProfile = function() {
+  var self = this;
+  return new Promise(function(success, error) {
+    var userProfile = localStorage.getItem();
```
Comment (JameelB) [inline packages/angularjs-auth-keycloak/lib/keycloakAuth.js]: Missing item key

File: packages/angularjs-auth-keycloak/lib/keycloakAuth.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      }).error(error);
+    }
+  });
+}
+
+KeycloakAuth.prototype.setListener = function() {
+  // NOTE: No need to listen to login/logout events for Keycloak as the login page is in a different page from the app itself.
+};
+
+KeycloakAuth.prototype.setKeycloakLib = function(keycloakLib) {
```
Comment (JameelB) [inline packages/angularjs-auth-keycloak/lib/keycloakAuth.js]: What will this be used for?

File: demo/portal/src/app/main.js [hunk 3]
```diff
@@ -24,7 +24,10 @@ logger.setLogger(new logger.ClientLogger(2));
 angular.module('app', [
   require('angular-ui-router'),
   require('angular-material'),
+  // Enables passport auth service to be used
   require('./passport'),
+  // Enables keycloak auth service to be used
+  require('./keycloak'),
```
Comment (JameelB) [inline demo/portal/src/app/main.js]: this should be commented out.

Comment (JameelB) [inline demo/portal/src/app/main.js]: this should be commented out

File: packages/angularjs-auth-keycloak/lib/keycloakAuth.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  this.keycloakLib = keycloakLib;
+};
+
+KeycloakAuth.prototype.login = function() {
+  if (this.keycloakLib) {
+    this.keycloakLib.login(arguments);
+  }
+}
+
+KeycloakAuth.prototype.logout = function() {
```
Comment (JameelB) [inline packages/angularjs-auth-keycloak/lib/keycloakAuth.js]: This should also clear localStorage

File: demo/mobile/src/app/app.js [hunk 5]
```diff
@@ -15,7 +15,9 @@ angular.module('wfm-mobile', [
   require('angular-ui-router'),
   require('angular-material'),
   // Enables passport auth service to be used
-  require('./passport'),
+  // require('./passport'),
```
Comment (JameelB) [inline demo/mobile/src/app/app.js]: it should use passport as default

File: demo/portal/src/app/keycloak/index.js [hunk 6]
```diff
@@ -1,3 +1,19 @@
 var keycloakConfig = require('../../config.json');
+var keycloakModule = require('@raincatcher/angularjs-auth-keycloak');
 
-require('@raincatcher/angularjs-auth-keycloak')('app', keycloakConfig.keycloak, keycloakConfig.keycloakInit);
+var MOBILE_AUTH_MODULE_ID = "wfm.auth.mobile.keycloak";
+
+var angularModule = angular.module(MOBILE_AUTH_MODULE_ID, []);
+var keycloakLib;
+if (window.navigator.onLine) {
+  keycloakLib = keycloakModule.init('wfm-mobile', angularModule,
```
Comment (JameelB) [inline demo/portal/src/app/keycloak/index.js]: The appname should be `app`

**Review by JameelB — APPROVED** [review]

**Comment by wtrocki** [comment]
Verified locally. 

- [x] When user is running app and it's not logged in application will notify him by dialog
- [x] When user is running app and it's logged in application will start
(Workorders will not appear - need to clarify that when running on the server) 
- [x] Logout when offline opens dialog
- [x] When logging to portal as unauthorized user redirect url for keycloak doesn't contain unauthorized path.

ping @JameelB @witmicko

**Pull request merged**